### ISSUE TITLE
v0.3 regression on OCaml < 4.07

### ISSUE THREAD

**Issue by sim642** [c-cube/seq#14]
I have some nightly CI builds that suddenly started failing (with no code changes!) and the only difference is seq 0.3 being installed instead of 0.2.2, so this seems to be the culprit. The regression is a runtime crash:
```
Fatal error: exception File "src/select_version.ml", line 11, characters 11-17: Assertion failed
```
which looks like the case here:
https://github.com/c-cube/seq/blob/7b957912666a3acb69e9e5e334367c7d58048aba/src/select_version.ml#L11

I'm confused by what `Select_version` is still for, it doesn't seem to be used for any build time version selection anymore, but it's still there, so I guess it gets implicitly linked into the library and unintentionally runs some code every time?

The regression didn't show in opam-repository CI because the latest released version of my package (goblint-cil) doesn't yet have a transitive dependency on seq, but the development version does (via yojson 2.0.0).
Also, I guess my use of `-linkall` is what reveals the issue (by running `Select_version`). Default linking would exclude the unused module.

**Comment by sim642** [comment]
Coincidentally, a very similar issue just appeared on a completely unrelated project as well: namely, dune's mdx support also uses `-linkall` and if seq 0.3 is installed, mdx crashes with the same exception.

**Comment by sim642** [comment]
The regression also now shows in the latest opam-health-check diff: http://check.ocamllabs.io/diff/1658498706-41f2345be04c54b77fe87c9100aef20218bcac77..1658671506-f1eef1a5b413443a841a24a1c7f655ef64772692. May of the now failing are with the same exception.

The same can be seen by grepping the build logs for `select_version.ml` on that run: http://check.ocamllabs.io/run/1658671506-f1eef1a5b413443a841a24a1c7f655ef64772692?comp=4.02&comp=4.03&comp=4.04&comp=4.05&comp=4.06&comp=4.07&comp=4.08&comp=4.09&comp=4.10&comp=4.11&comp=4.12&comp=4.13&comp=4.14&comp=4.14%2Balpha-repo&comp=5.0&comp=5.0%2Balpha-repo&available=4.02&available=4.03&available=4.04&available=4.05&available=4.06&available=4.07&available=4.08&available=4.09&available=4.10&available=4.11&available=4.12&available=4.13&available=4.14&available=4.14%2Balpha-repo&available=5.0&available=5.0%2Balpha-repo&maintainers=&logsearch=select_version.ml&logsearch_comp=4.06.

Surprisingly fewer than I expected based on my personal experience with this crash.
